native/jni/magiskhide/proc_monitor.cpp

```diff
@@ -106,6 +146,209 @@ static void hide_daemon(int pid) {
 	_exit(0);
 }
 
+/*
+ * Bionic's atoi runs through strtol() and fault-tolerence checkings.
+ * Since we don't need it, use our own implementation of atoi()
+ * for faster conversion.
+ */
+static inline int fast_atoi(const char *str) {
+	int val = 0;
+
+	while (*str)
+		val = val * 10 + (*str++ - '0');
+
+	return val;
+}
+
+// Leave /proc fd opened as we're going to read from it repeatedly
+static DIR *dfd;
+// Use unordered map with pid and namespace inode number to avoid time-consuming GC
+static unordered_map<int, uint64_t> pid_ns_map;
+// Use set for slow insertion but fast searching(which we'd encounter a lot more)
+static set<uid_t> hide_uid;
+// Treat GMS separately as we're only interested in one component
+static int gms_uid = -1;
+
+static void detect_new_processes() {
+	struct dirent *dp;
+	struct stat ns, pns;
+	int pid, ppid;
+	bool hide;
+	uid_t uid;
+	unordered_map<int, uint64_t>::const_iterator pos;
+
+	// Iterate through /proc and get a process that reads the target APK
+	rewinddir(dfd);
+	while ((dp = readdir(dfd))) {
+		if (!isdigit(dp->d_name[0]))
+			continue;
+
+		// dp->d_name is now the pid
+		pid = fast_atoi(dp->d_name);
+
+		// We're only interested in PIDs > 1000
+		if (pid <= 1000)
+			continue;
+
+		uid = get_uid(pid) % 100000; // Handle multiuser
+		if (hide_uid.find(uid) != hide_uid.end()) {
+			// Make sure our target is alive
+			if ((ppid = parse_ppid(pid)) < 0 || read_ns(ppid, &pns) || read_ns(pid, &ns))
+				continue;
+
+			// Check if it's a process we haven't already hijacked
+			hide = false;
+			pos = pid_ns_map.find(pid);
+			if (pos == pid_ns_map.end()) {
+				hide = true;
+				pid_ns_map.insert(pair<int, uint64_t>(pid, ns.st_ino));
+			} else if (pos->second != ns.st_ino) {
+				hide = true;
```

Comment:
Why true? We don't have to repeat the unmount again.

Reply:
This is intentional.

inotify will report events even when an app didn't actually launch, so to prevent repetitive unmounts on an already hijacked process, the list of hijacked processes will be recorded in pid_ns_map.
In this case(pos->second != ns.st_ino), the pid was previously hijacked, but is now a different one because the previous one is terminated. Linux reuses pid numbers after all. So re-unmounting is necessary and the list is updated accordingly.

Reply:
Memory usage will continue to increase

Reply:
Theoretic maximum in absolute worst case is 8*32K as pid number is limited to 32K.
I chose to leave it that way without GC as traversing through the map on every detect_new_processes() is very inefficient.

But on a second thought, GC'ing the map upon update_apk_list() seems like a good idea and it'll even solve the map getting to the theoretical max size on extremely long uptime.

Will send another patch soon.